src/Compilers/CSharp/Portable/Symbols/MemberSymbolExtensions.cs

```diff
@@ -407,8 +407,27 @@ internal static bool IsDefaultValueTypeConstructor(this NamedTypeSymbol type, Co
             {
                 return false;
             }
-            var constructor = type.InstanceConstructors.SingleOrDefault(m => m.ParameterCount == 0);
-            return constructor?.IsDefaultValueTypeConstructor(requireZeroInit: true) == true;
+
+            bool foundParameterlessCtor = false;
+            bool result = false;
+            foreach (var constructor in type.InstanceConstructors)
+            {
+                if (constructor.ParameterCount != 0)
+                {
+                    continue;
+                }
+
+                if (foundParameterlessCtor)
+                {
+                    // finding more than one parameterless constructor is an error scenario (reported elsewhere)
+                    return false;
+                }
+
+                foundParameterlessCtor = true;
+                result = constructor.IsDefaultValueTypeConstructor(requireZeroInit: true);
+            }
+
+            return result;
```

Comment:
I was just perusing open pull requests. Apologies for dropping in randomly and offering my point of view, but it took me awhile to figure out what this code is trying to do.
I wonder whether the intent is better captured by code like
```
var parameterlessCtors = type.InstanceConstructors.Where(ctor => ctor.ParameterCount == 0).ToList();
switch (parameterlessCtors.Count()) {
    case 0: return false;
    case 1: return parameterlessCtors.Single().IsDefaultValueTypeConstructor(requireZeroInit: true);
    // finding more than one parameterless constructor is an error scenario (reported elsewhere)
    default: return false;
}
```

Reply:
I'll add a comment for clarity, but the reason for this loop is to not allocate because of a possible (but uncommon) scenario.